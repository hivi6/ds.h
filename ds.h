// ==================================================
// DEFINITIONS
// ==================================================

#ifndef DS_H
#define DS_H

// --------------------------------------------------
// error
// --------------------------------------------------

enum
{
    DS_NO_ERR = 0,
    DS_NULL_ERR,
    DS_MALLOC_ERR,
    DS_RANGE_ERR,
    DS_SIZE_ERR,
    DS_EMPTY_ERR,
    DS_NUM_OF_ERRORS
};

const char *ds_error(int errno);

// --------------------------------------------------
// vector
// --------------------------------------------------

typedef struct
{
    int cap; 	  // capacity of the vector
    int cnt;      // number of items in the vector
    void **items; // array of items
    int *sizes;   // size of each item
} vec_t;

int vec_init(vec_t *vec);
int vec_free(vec_t *vec);
int vec_append(vec_t *vec, void *item, int size);
int vec_pop(vec_t *vec);
int vec_get(vec_t *vec, int index, void *item, int size);
int vec_set(vec_t *vec, int index, void *item, int size);

// --------------------------------------------------
// string builder
// --------------------------------------------------

typedef struct
{
    vec_t chars;
} sbuilder_t;

int sbuilder_init(sbuilder_t *sbuilder);
int sbuilder_free(sbuilder_t *sbuilder);
int sbuilder_append(sbuilder_t *sbuilder, const char *str);
int sbuilder_appendc(sbuilder_t *sbuilder, char ch);
int sbuilder_pop(sbuilder_t *sbuilder);
int sbuilder_get(sbuilder_t *sbuilder, int index, char *ch);
int sbuilder_set(sbuilder_t *sbuilder, int index, char ch);
int sbuilder_build(sbuilder_t *sbuilder, char **str);

#endif // DS_H

// ==================================================
// IMPLEMENTATIONS
// ==================================================

#ifdef DS_IMPLEMENTATION

#include <stdlib.h>
#include <string.h>

// --------------------------------------------------
// error
// --------------------------------------------------

const char *ds_error(int errno)
{
    static const char *msg[] = {
	"No error", // DS_NO_ERR = 0,
	"NULL argument error", // DS_NULL_ERR,
	"Malloc error", // DS_MALLOC_ERR,
	"Index out of range error", // DS_RANGE_ERR,
	"Mismatch size error", // DS_SIZE_ERR,
	"Empty vector error", // DS_EMPTY_ERR,
    };

    if (errno >= DS_NUM_OF_ERRORS)
    {
	return "What is this errno?";
    }
    return msg[errno];
}

// --------------------------------------------------
// vector
// --------------------------------------------------

int vec_init(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;

    vec->cap = 0;
    vec->cnt = 0;
    vec->items = NULL;
    vec->sizes = NULL;
    return DS_NO_ERR;
}

int vec_free(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;

    for (int i = 0; i < vec->cnt; i++) free(vec->items[i]);
    free(vec->items);
    free(vec->sizes);

    return vec_init(vec);
}

int vec_append(vec_t *vec, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;

    if (vec->cnt == vec->cap)
    {
	int new_cap = (vec->cap < 8 ? 8 : 2 * vec->cap);
	void **new_items = malloc(new_cap * sizeof(void *));
	int *new_sizes = malloc(new_cap * sizeof(int));
	if (new_items == NULL || new_sizes == NULL) return DS_MALLOC_ERR;

	for (int i = 0; i < vec->cap; i++)
	{
	    new_items[i] = vec->items[i];
	    new_sizes[i] = vec->sizes[i];
	}
	free(vec->items);
	free(vec->sizes);

	vec->cap = new_cap;
	vec->items = new_items;
	vec->sizes = new_sizes;
    }

    char *temp = malloc(sizeof(char) * size);
    if (temp == NULL) return DS_MALLOC_ERR;
    memcpy(temp, item, size);

    vec->items[vec->cnt] = temp;
    vec->sizes[vec->cnt] = size;
    vec->cnt++;
    return DS_NO_ERR;
}

int vec_pop(vec_t *vec)
{
    if (vec == NULL) return DS_NULL_ERR;
    if (vec->cnt == 0) return DS_EMPTY_ERR;

    free(vec->items[vec->cnt - 1]);
    vec->cnt--;
    return DS_NO_ERR;
}

int vec_get(vec_t *vec, int index, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;
    if (index < 0 || index >= vec->cnt) return DS_RANGE_ERR;
    if (vec->sizes[index] != size) return DS_SIZE_ERR;

    memcpy(item, vec->items[index], size);
    return DS_NO_ERR;
}

int vec_set(vec_t *vec, int index, void *item, int size)
{
    if (vec == NULL || item == NULL) return DS_NULL_ERR;
    if (index < 0 || index >= vec->cnt) return DS_RANGE_ERR;

    char *temp = malloc(sizeof(char) * size);
    if (temp == NULL) return DS_MALLOC_ERR;
    memcpy(temp, item, size);

    free(vec->items[index]);
    vec->items[index] = temp;
    vec->sizes[index] = size;
    return DS_NO_ERR;
}

// --------------------------------------------------
// string builder
// --------------------------------------------------

int sbuilder_init(sbuilder_t *sbuilder)
{
    if (sbuilder == NULL) return DS_NULL_ERR;
    return vec_init(&sbuilder->chars);
}

int sbuilder_free(sbuilder_t *sbuilder)
{
    if (sbuilder == NULL) return DS_NULL_ERR;
    return vec_free(&sbuilder->chars);
}

int sbuilder_append(sbuilder_t *sbuilder, const char *str)
{
    if (sbuilder == NULL || str == NULL) return DS_NULL_ERR;
    for (int i = 0; str[i]; i++)
    {
	int errno = vec_append(&sbuilder->chars, (void*)str + i, sizeof(char));
	if (errno) return errno;
    }
    return DS_NO_ERR;
}

int sbuilder_appendc(sbuilder_t *sbuilder, char ch)
{
    if (sbuilder == NULL) return DS_NULL_ERR;
    return vec_append(&sbuilder->chars, &ch, sizeof(ch));
}

int sbuilder_pop(sbuilder_t *sbuilder)
{
    if (sbuilder == NULL) return DS_NULL_ERR;
    return vec_pop(&sbuilder->chars);
}

int sbuilder_get(sbuilder_t *sbuilder, int index, char *ch)
{
    if (sbuilder == NULL || ch == NULL) return DS_NULL_ERR;
    return vec_get(&sbuilder->chars, index, ch, sizeof(*ch));
}

int sbuilder_set(sbuilder_t *sbuilder, int index, char ch)
{
    if (sbuilder == NULL) return DS_NULL_ERR;
    return vec_set(&sbuilder->chars, index, &ch, sizeof(ch));
}

int sbuilder_build(sbuilder_t *sbuilder, char **str)
{
    if (sbuilder == NULL || str == NULL) return DS_NULL_ERR;

    int len = sbuilder->chars.cnt;
    char *temp = malloc(sizeof(char) * len + 1);
    temp[len] = 0;
    for (int i = 0; i < len; i++)
    {	
	int errno = sbuilder_get(sbuilder, i, temp + i);
	if (errno) return errno;
    }

    *str = temp;
    return DS_NO_ERR;
}

#endif // DS_IMPLEMENTATION
