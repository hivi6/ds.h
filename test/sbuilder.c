#include <stdio.h>

#define DS_IMPLEMENTATION
#include "ds.h"

int main()
{
    sbuilder_t sbuilder;
    int errno = sbuilder_init(&sbuilder);
    printf("sbuilder_init() | ERROR: %s\n", ds_error(errno));

    for (int i = 0; i < 26; i++)
    {
	errno = sbuilder_appendc(&sbuilder, 'a' + i);
	printf("sbuilder_appendc(%c) | LEN: %d | ERROR: %s | STR: ", 'a' + i, 
	       sbuilder.chars.cnt, ds_error(errno));
	char *str;
	sbuilder_build(&sbuilder, &str);
	printf("%s\n", str);
	free(str);
    }

    for (int i = 0; sbuilder.chars.cnt; i++)
    {
	for (int j = 0; j < 5; j++)
	{
	    errno = sbuilder_pop(&sbuilder);
	    printf("sbuilder_pop() | LEN: %d | ERROR: %s | STR: ", sbuilder.chars.cnt, 
		   ds_error(errno));
	    char *str;
	    sbuilder_build(&sbuilder, &str);
	    printf("%s\n", str);
	    free(str);
	}
    }

    const char *cstr = "this is a string!";
    sbuilder_append(&sbuilder, cstr);
    char *str;
    errno = sbuilder_build(&sbuilder, &str);
    printf("sbuilder_append(%s) | LEN: %d | ERROR: %s | STR: %s\n", cstr, 
	   sbuilder.chars.cnt, ds_error(errno), str);
    free(str);

    for (int i = 0; sbuilder.chars.cnt; i++)
    {
	for (int j = 0; j < 5; j++)
	{
	    errno = sbuilder_pop(&sbuilder);
	    printf("sbuilder_pop() | LEN: %d | ERROR: %s | STR: ", sbuilder.chars.cnt, 
		   ds_error(errno));
	    char *str;
	    sbuilder_build(&sbuilder, &str);
	    printf("%s\n", str);
	    free(str);
	}
    }

    sbuilder_free(&sbuilder);
    printf("sbuilder_free() | ERROR: %s\n", ds_error(errno));

    return 0;
}
