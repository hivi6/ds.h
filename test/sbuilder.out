sbuilder_init() | ERROR: No error
sbuilder_appendc(a) | LEN: 1 | ERROR: No error | STR: a
sbuilder_appendc(b) | LEN: 2 | ERROR: No error | STR: ab
sbuilder_appendc(c) | LEN: 3 | ERROR: No error | STR: abc
sbuilder_appendc(d) | LEN: 4 | ERROR: No error | STR: abcd
sbuilder_appendc(e) | LEN: 5 | ERROR: No error | STR: abcde
sbuilder_appendc(f) | LEN: 6 | ERROR: No error | STR: abcdef
sbuilder_appendc(g) | LEN: 7 | ERROR: No error | STR: abcdefg
sbuilder_appendc(h) | LEN: 8 | ERROR: No error | STR: abcdefgh
sbuilder_appendc(i) | LEN: 9 | ERROR: No error | STR: abcdefghi
sbuilder_appendc(j) | LEN: 10 | ERROR: No error | STR: abcdefghij
sbuilder_appendc(k) | LEN: 11 | ERROR: No error | STR: abcdefghijk
sbuilder_appendc(l) | LEN: 12 | ERROR: No error | STR: abcdefghijkl
sbuilder_appendc(m) | LEN: 13 | ERROR: No error | STR: abcdefghijklm
sbuilder_appendc(n) | LEN: 14 | ERROR: No error | STR: abcdefghijklmn
sbuilder_appendc(o) | LEN: 15 | ERROR: No error | STR: abcdefghijklmno
sbuilder_appendc(p) | LEN: 16 | ERROR: No error | STR: abcdefghijklmnop
sbuilder_appendc(q) | LEN: 17 | ERROR: No error | STR: abcdefghijklmnopq
sbuilder_appendc(r) | LEN: 18 | ERROR: No error | STR: abcdefghijklmnopqr
sbuilder_appendc(s) | LEN: 19 | ERROR: No error | STR: abcdefghijklmnopqrs
sbuilder_appendc(t) | LEN: 20 | ERROR: No error | STR: abcdefghijklmnopqrst
sbuilder_appendc(u) | LEN: 21 | ERROR: No error | STR: abcdefghijklmnopqrstu
sbuilder_appendc(v) | LEN: 22 | ERROR: No error | STR: abcdefghijklmnopqrstuv
sbuilder_appendc(w) | LEN: 23 | ERROR: No error | STR: abcdefghijklmnopqrstuvw
sbuilder_appendc(x) | LEN: 24 | ERROR: No error | STR: abcdefghijklmnopqrstuvwx
sbuilder_appendc(y) | LEN: 25 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxy
sbuilder_appendc(z) | LEN: 26 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxyz
sbuilder_pop() | LEN: 25 | ERROR: No error | STR: abcdefghijklmnopqrstuvwxy
sbuilder_pop() | LEN: 24 | ERROR: No error | STR: abcdefghijklmnopqrstuvwx
sbuilder_pop() | LEN: 23 | ERROR: No error | STR: abcdefghijklmnopqrstuvw
sbuilder_pop() | LEN: 22 | ERROR: No error | STR: abcdefghijklmnopqrstuv
sbuilder_pop() | LEN: 21 | ERROR: No error | STR: abcdefghijklmnopqrstu
sbuilder_pop() | LEN: 20 | ERROR: No error | STR: abcdefghijklmnopqrst
sbuilder_pop() | LEN: 19 | ERROR: No error | STR: abcdefghijklmnopqrs
sbuilder_pop() | LEN: 18 | ERROR: No error | STR: abcdefghijklmnopqr
sbuilder_pop() | LEN: 17 | ERROR: No error | STR: abcdefghijklmnopq
sbuilder_pop() | LEN: 16 | ERROR: No error | STR: abcdefghijklmnop
sbuilder_pop() | LEN: 15 | ERROR: No error | STR: abcdefghijklmno
sbuilder_pop() | LEN: 14 | ERROR: No error | STR: abcdefghijklmn
sbuilder_pop() | LEN: 13 | ERROR: No error | STR: abcdefghijklm
sbuilder_pop() | LEN: 12 | ERROR: No error | STR: abcdefghijkl
sbuilder_pop() | LEN: 11 | ERROR: No error | STR: abcdefghijk
sbuilder_pop() | LEN: 10 | ERROR: No error | STR: abcdefghij
sbuilder_pop() | LEN: 9 | ERROR: No error | STR: abcdefghi
sbuilder_pop() | LEN: 8 | ERROR: No error | STR: abcdefgh
sbuilder_pop() | LEN: 7 | ERROR: No error | STR: abcdefg
sbuilder_pop() | LEN: 6 | ERROR: No error | STR: abcdef
sbuilder_pop() | LEN: 5 | ERROR: No error | STR: abcde
sbuilder_pop() | LEN: 4 | ERROR: No error | STR: abcd
sbuilder_pop() | LEN: 3 | ERROR: No error | STR: abc
sbuilder_pop() | LEN: 2 | ERROR: No error | STR: ab
sbuilder_pop() | LEN: 1 | ERROR: No error | STR: a
sbuilder_pop() | LEN: 0 | ERROR: No error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_append(this is a string!) | LEN: 17 | ERROR: No error | STR: this is a string!
sbuilder_pop() | LEN: 16 | ERROR: No error | STR: this is a string
sbuilder_pop() | LEN: 15 | ERROR: No error | STR: this is a strin
sbuilder_pop() | LEN: 14 | ERROR: No error | STR: this is a stri
sbuilder_pop() | LEN: 13 | ERROR: No error | STR: this is a str
sbuilder_pop() | LEN: 12 | ERROR: No error | STR: this is a st
sbuilder_pop() | LEN: 11 | ERROR: No error | STR: this is a s
sbuilder_pop() | LEN: 10 | ERROR: No error | STR: this is a 
sbuilder_pop() | LEN: 9 | ERROR: No error | STR: this is a
sbuilder_pop() | LEN: 8 | ERROR: No error | STR: this is 
sbuilder_pop() | LEN: 7 | ERROR: No error | STR: this is
sbuilder_pop() | LEN: 6 | ERROR: No error | STR: this i
sbuilder_pop() | LEN: 5 | ERROR: No error | STR: this 
sbuilder_pop() | LEN: 4 | ERROR: No error | STR: this
sbuilder_pop() | LEN: 3 | ERROR: No error | STR: thi
sbuilder_pop() | LEN: 2 | ERROR: No error | STR: th
sbuilder_pop() | LEN: 1 | ERROR: No error | STR: t
sbuilder_pop() | LEN: 0 | ERROR: No error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_pop() | LEN: 0 | ERROR: Empty vector error | STR: 
sbuilder_free() | ERROR: Empty vector error
